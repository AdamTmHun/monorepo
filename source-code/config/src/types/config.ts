import type * as ast from "@inlang/ast";
import type { LanguageCode } from "../../../i18n-utilities/dist/index.js";
import type { FsPromisesSubset } from "./fsPromisesSubset.js";
import type { DocumentSelector } from "vscode";

/**
 * Inlang config schema.
 */
export type Config = {
	/**
	 * The language that other languages are validated against.
	 *
	 * In most cases, the reference language is `en` (English).
	 */
	referenceLanguage: LanguageCode;
	/**
	 * Languages of this project.
	 *
	 * The languages must include the reference language itself.
	 */
	languages: LanguageCode[];
	readResource: (args: {
		fs: FsPromisesSubset;
		languageCode: LanguageCode;
		ast: typeof ast;
	}) => Promise<ast.Resource>;
	writeResource: (args: {
		fs: FsPromisesSubset;
		resource: ast.Resource;
		languageCode: LanguageCode;
	}) => Promise<void>;
	ideExtension: {
		extractMessageReplacementOptions: (args: { id: string }) => string[];
		/**
		 * An array of VSCode DocumentSelectors.
		 *
		 * The document selectors specify for which files/programming languages
		 * (typescript, svelte, etc.) the extension should be activated.
		 *
		 * See https://code.visualstudio.com/api/references/document-selector
		 */
		documentSelectors: DocumentSelector[];
	};
};
