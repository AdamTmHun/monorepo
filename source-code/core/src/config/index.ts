export { Config } from "./schema.js";
