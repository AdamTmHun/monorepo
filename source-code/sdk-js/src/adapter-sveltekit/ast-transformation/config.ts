import { readFile } from "node:fs/promises"
import { initConfig } from "../../config/index.js"
import { stat } from "node:fs/promises"
import type { CompilerOptions } from "typescript"
import { parse } from "json5"

export const doesPathExist = async (path: string) => !!(await stat(path).catch(() => undefined))

export type TransformConfig = {
	languageInUrl: boolean
	isStatic: boolean
	srcFolder: string
	rootRoutesFolder: string
	hasAlreadyBeenInitialized: boolean
	tsCompilerOptions: CompilerOptions
	sourceFileName?: string
	sourceMapName?: string
}

let cachedConfig: TransformConfig | undefined

const cwd = process.cwd()

export const getConfig = async (): Promise<TransformConfig> => {
	if (cachedConfig) return cachedConfig

	const inlangConfig = await initConfig()
	const svelteConfig = await readFile("svelte.config.js", "utf-8")
	const tsConfig = await readFile("tsconfig.json", "utf-8")
	const tsCompilerOptions = parse(tsConfig).compilerOptions
	// TODO get JSON 5 parser to parse typescript config and only pass property compilerOptions

	const languageInUrl =
		inlangConfig?.sdk?.languageNegotiation?.strategies?.some(({ type }) => type === "url") || false
	const isStatic = svelteConfig.includes("@sveltejs/adapter-static") // TODO: static means if `prerender` is set to true at the root

	const srcFolder = cwd + "/src"
	const rootRoutesFolder = srcFolder + "/routes/(app)" + (languageInUrl ? "/[lang]" : "")

	const hasAlreadyBeenInitialized = await doesPathExist(rootRoutesFolder)

	return (cachedConfig = {
		languageInUrl,
		isStatic,
		srcFolder,
		rootRoutesFolder,
		hasAlreadyBeenInitialized,
		tsCompilerOptions,
	})
}

export const resetConfig = () => (cachedConfig = undefined)
