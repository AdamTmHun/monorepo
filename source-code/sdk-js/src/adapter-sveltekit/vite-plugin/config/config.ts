import { ConfigPathNotFoundError, createInlang, tryCatch, LanguageTag, Message } from "@inlang/app"
import { InlangSdkException } from "../exceptions.js"
import path, { resolve } from "node:path"
import type { Config as SvelteConfig } from "@sveltejs/kit"
import type { SdkConfig } from '@inlang/sdk-js-plugin'
import { getSvelteKitVersion } from './utils/getSvelteKitVersion.js'
import { shouldContentBePrerendered } from './utils/shouldContentBePrerendered.js'
import { updateSdkModuleVersion } from './utils/updateSdkModuleVersion.js'
import { getSettings } from './utils/getSettings.js'
import { createBasicInlangConfig } from './utils/createBasicInlangConfig.js'
import { createDemoResourcesIfNoMessagesExistYet } from './utils/createDemoResourcesIfNoMessagesExistYet.js'
import { doesPathExist } from './utils/utils.js'
import { getNodeishFs } from './utils/getNodeishFs.js'

type VersionString = `${number}.${number}.${number}${string}`

export type TransformConfig = {
	debug: boolean

	sourceLanguageTag: LanguageTag
	languageTags: LanguageTag[]
	messages: () => Message[]

	cwdFolderPath: string
	options: {
		rootRoutesFolder: string
		languageInUrl: boolean
		isStatic: boolean
		resourcesCache: SdkConfig["resources"]["cache"]
		excludedRoutes: SdkConfig["routing"]["exclude"]
	}

	svelteKit: {
		version: VersionString | undefined
		usesTypeScript: boolean
		files: {
			appTemplate: string
			routes: string
			serverHooks: string
		}
	}
}

export const PATH_TO_CWD = process.cwd()
export const PATH_TO_INLANG_CONFIG = resolve(PATH_TO_CWD, "./inlang.config.json")

let transformConfig: Promise<TransformConfig> | undefined = undefined

export const initTransformConfig = async (): Promise<TransformConfig> => {
	if (transformConfig) return transformConfig

	// eslint-disable-next-line no-async-promise-executor
	return (transformConfig = new Promise<TransformConfig>(async (resolve, reject) => {
		const { data: inlang, error: createInlangError } = await tryCatch(async () =>
			createInlang({ nodeishFs: await getNodeishFs(), configPath: PATH_TO_INLANG_CONFIG })
		)
		if (createInlangError) {
			if (createInlangError instanceof ConfigPathNotFoundError) {
				await createBasicInlangConfig()
				transformConfig = undefined
				return resolve(initTransformConfig())
			}

			return reject(createInlangError)
		}

		if (inlang.errors.length) {
			return reject(inlang.errors)
		}

		await updateSdkModuleVersion(inlang)
		await createDemoResourcesIfNoMessagesExistYet(inlang)

		const settings = getSettings(inlang)
		if (!settings) {
			transformConfig = undefined
			return resolve(initTransformConfig())
		}

		const { default: svelteConfig } = (await import(/* @vite-ignore */
			/*pathToFileURL*/(path.resolve(PATH_TO_CWD, "svelte.config.js")).toString()
		).catch((error: unknown) => {
			throw new InlangSdkException('Could not find svelte.config.js file.', error as Error)
		})) as { default: SvelteConfig }

		const files = {
			appTemplate: path.resolve(
				PATH_TO_CWD,
				svelteConfig.kit?.files?.appTemplate || path.resolve("src", "app.html"),
			),
			routes: path.resolve(
				PATH_TO_CWD,
				svelteConfig.kit?.files?.routes || path.resolve("src", "routes"),
			),
			serverHooks: path.resolve(
				PATH_TO_CWD,
				svelteConfig.kit?.files?.hooks?.server || path.resolve("src", "hooks.server"),
			),
		}

		const languageInUrl =
			settings.languageNegotiation.strategies.some(({ type }) => type === "url") ||
			false

		const rootRoutesFolder = path.resolve(files.routes, languageInUrl ? "[lang]" : "")
		const isStatic =
			(await shouldContentBePrerendered(files.routes)) ||
			(await shouldContentBePrerendered(rootRoutesFolder))

		const usesTypeScript = await doesPathExist(path.resolve(PATH_TO_CWD, "tsconfig.json"))

		const svelteKitVersion = await getSvelteKitVersion()

		resolve({
			debug: settings.debug,

			sourceLanguageTag: inlang.config().sourceLanguageTag,
			languageTags: inlang.config().languageTags,
			messages: () => inlang.query.messages.getAll(),

			cwdFolderPath: PATH_TO_CWD,
			options: {
				rootRoutesFolder,
				languageInUrl,
				isStatic,
				resourcesCache: settings.resources.cache,
				excludedRoutes: settings.routing.exclude,
			},

			svelteKit: {
				version: svelteKitVersion,
				usesTypeScript,
				files,
			},
		})
	}))
}

export const resetTransformConfig = () => (transformConfig = undefined)
