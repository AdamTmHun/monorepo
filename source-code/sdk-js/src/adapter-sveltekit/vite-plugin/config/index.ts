export { type TransformConfig, initTransformConfig, resetTransformConfig } from './config.js'