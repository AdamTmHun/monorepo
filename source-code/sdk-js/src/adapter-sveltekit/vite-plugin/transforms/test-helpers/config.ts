import type { TransformConfig } from "../../config.js"

export const baseTestConfig: TransformConfig = {
	isStatic: false,
	languageInUrl: false,
	srcFolder: "",
	rootRoutesFolder: "",
	sourceFileName: "",
	sourceMapName: "",
	isTypeScriptProject: false
}
