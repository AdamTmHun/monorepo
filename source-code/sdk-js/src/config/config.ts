import type { NodeishFilesystemSubset } from "@inlang/app"
import type { InlangConfig } from "@inlang/app"
import type { SdkConfig } from "@inlang/sdk-js-plugin"
import { dedent } from "ts-dedent"
import { InlangSdkException } from "../adapter-sveltekit/vite-plugin/exceptions.js"

export type InlangConfigWithSdkProps = InlangConfig & {
	sdk: SdkConfig
}

// @ts-ignore
export const initInlangEnvironment = async (): Promise<InlangEnvironment> => {
	const fs = await import("node:fs/promises").catch(
		() =>
			new Proxy({} as NodeishFilesystemSubset, {
				get: (target, key) => {
					if (key === "then") return Promise.resolve(target)

					return () => {
						throw new InlangSdkException(
							"`node:fs/promises` is not available in the current environment",
						)
					}
				},
			}),
	)

	return {
		$fs: fs,
		// @ts-ignore
		$import: initialize$import({
			fs,
			// @ts-ignore
			fetch: async (...args) =>
				// @ts-ignore
				await fetch(...args).catch((error) => {
					// TODO: create an issue
					if (
						error instanceof TypeError &&
						(error.cause as any)?.code === "UND_ERR_CONNECT_TIMEOUT"
					) {
						throw new InlangSdkException(
							dedent`
								Node.js failed to resolve the URL. This can happen sometimes during development.
								Usually restarting the server helps.
							`,
							error,
						)
					}

					throw new InlangSdkException(`Failed to fetch from (${args[0]})`, error)
				}),
		}),
	}
}

export const initConfig = async (module: any) => {
	if (!module) {
		throw new InlangSdkException("could not read `inlang.config.js`")
	}
	const env = await initInlangEnvironment()

	// @ts-ignore
	return setupConfig({ module, env })
}
