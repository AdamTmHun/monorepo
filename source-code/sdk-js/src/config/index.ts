import { InlangConfigModule, setupConfig } from "@inlang/core/config"
import { initialize$import } from "@inlang/core/environment"
import fs from "node:fs/promises"
import { resolve } from "node:path"

// eslint-disable-next-line @typescript-eslint/no-empty-function
async function readInlangConfig() {
	const env = {
		$fs: fs,
		$import: initialize$import({
			fs,
			fetch,
		}),
	}

	const module = (await import(
		/* @vite-ignore */ resolve(process.cwd(), "./inlang.config.js")
	)) as InlangConfigModule

	const config = await setupConfig({ module, env })

	return config
}

export const initConfig = async () => {
	const config = await readInlangConfig()
	if (!config) return undefined

	// TODO: validate sdk related once we add entries to it

	// set defaults
	// TODO: set defaults for the IDE extension
	// TODO: setting defaults needs to happen inside `defineConfig` so the IDE extension can use them
	config.sdk = {
		...config.sdk,
		languageNegotiation: {
			strict: false,
			...config.sdk?.languageNegotiation,
			strategies: config.sdk?.languageNegotiation?.strategies || [
				{ type: "url" },
				{ type: "accept-language-header" },
				{ type: "navigator" },
			],
		},
	}

	return config
}
