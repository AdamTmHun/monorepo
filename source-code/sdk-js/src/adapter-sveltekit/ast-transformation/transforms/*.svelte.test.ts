import { describe, it } from "vitest"
import { transformSvelte } from "./*.svelte.js"
import { baseTestConfig } from "./test-helpers/config.js"
import type { TransformConfig } from "../config.js"
import { readFileSync } from "node:fs"

const testSvelteFile = readFileSync(__dirname + "/test-helpers/test.svelte").toString()
describe("transformSvelte", () => {
	it("languageInUrl is true", async ({ expect }) => {
		const config: TransformConfig = {
			...baseTestConfig,
			languageInUrl: false,
			tsCompilerOptions: {
				allowJs: true,
				checkJs: true,
				esModuleInterop: true,
				forceConsistentCasingInFileNames: true,
				resolveJsonModule: true,
				skipLibCheck: true,
				sourceMap: true,
				strict: true,
				moduleResolution: "nodenext",
			},
			sourceFileName: "test.svelte",
			sourceMapName: "test.svelte.js",
		}
		const code = await transformSvelte(config, testSvelteFile)
		expect(code).toMatchSnapshot()
	})

	it("languageInUrl is false", async ({ expect }) => {
		const code = await transformSvelte(baseTestConfig, testSvelteFile)
		expect(code).toMatchSnapshot()
	})
})

// NOTES
// - Can merge imports of
//     - import { getRuntimeFromContext } from "@inlang/sdk-js/adapter-sveltekit/client/not-reactive";
//     - import { getRuntimeFromContext } from "@inlang/sdk-js/adapter-sveltekit/client/reactive";
// - Removes the i in "import { i, ... } from '@inlang/sdk-js'" or the complete import if no other module is imported.
// - Adds "const { i } = getRuntimeFromContext()" or "const { i: ... } = getRuntimFromContext()" if import aliases are used
// - Allows import aliasing of "import {i as ...} from '@inlang/sdk-js"
