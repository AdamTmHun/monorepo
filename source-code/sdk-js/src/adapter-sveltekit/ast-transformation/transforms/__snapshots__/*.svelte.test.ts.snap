// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transformSvelte > languageInUrl is false 1`] = `
"<script lang=\\"ts\\" context=\\"module\\"></script>

<script lang=\\"ts\\">import { getRuntimeFromContext } from \\"@inlang/sdk-js/adapter-sveltekit/client/reactive\\";

const {
  i: iStore,
  language: iLanguage
} = getRuntimeFromContext();

export let prop;
const blue = $iStore;
const green = $iLanguage;
console.log(blue);</script>

<style>
    .red {
        color: red;
    }
</style>

<h1 class=\\"red\\">{prop}</h1>
<h1 class=\\"red\\">{$iStore}</h1>
<h1 class=\\"red\\">{$iLanguage}</h1>
"
`;

exports[`transformSvelte > languageInUrl is true 1`] = `
"<script lang=\\"ts\\" context=\\"module\\"></script>

<script lang=\\"ts\\">import { getRuntimeFromContext } from \\"@inlang/sdk-js/adapter-sveltekit/client/reactive\\";

const {
  i: iStore,
  language: iLanguage
} = getRuntimeFromContext();

export let prop;
const blue = $iStore;
const green = $iLanguage;
console.log(blue);</script>

<style>
    .red {
        color: red;
    }
</style>

<h1 class=\\"red\\">{prop}</h1>
<h1 class=\\"red\\">{$iStore}</h1>
<h1 class=\\"red\\">{$iLanguage}</h1>
"
`;
